tests/posix/mqueue/src/mqueue_test.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <zephyr.h>
+#include <misc/printk.h>
+#include <mqueue.h>
+#include <pthread.h>
+
+#define N_THR 2
+#define STACKSZ 1024
+#define SENDER_THREAD 0
+#define RECEIVER_THREAD 1
+#define MESSAGE_SIZE	16
+#define MESG_COUNT_PERMQ	16
+
+
+K_THREAD_STACK_ARRAY_DEFINE(stacks, N_THR, STACKSZ);
+
+char queue[16] = "server";
+char send_data[MESSAGE_SIZE] = "timed data send";
+
+void *sender_thread(void *p1)
+{
+	mqd_t mqd;
+	struct timespec curtime;
+
+	mqd = mq_open(queue, O_WRONLY);
+	clock_gettime(CLOCK_MONOTONIC, &curtime);
+	curtime.tv_sec += 1;
+	zassert_false(mq_timedsend(mqd, send_data, MESSAGE_SIZE, 0, &curtime),
+		      "Not able to send message in timer\n");
+	usleep(USEC_PER_MSEC);
+	zassert_false(mq_close(mqd),
+		      "unable to close message queue descriptor.\n");
+	pthread_exit(p1);
+	return NULL;
+}
+
+
+void *receiver_thread(void *p1)
+{
+	mqd_t mqd;
+	char rec_data[MESSAGE_SIZE];
+	struct timespec curtime;
+
+	mqd = mq_open(queue, O_RDONLY);
+	clock_gettime(CLOCK_MONOTONIC, &curtime);
+	curtime.tv_sec += 1;
+	mq_timedreceive(mqd, rec_data, MESSAGE_SIZE, 0, &curtime);
+	zassert_false(strcmp(rec_data, send_data), "Error in data reception\n");
+	usleep(USEC_PER_MSEC);
+	zassert_false(mq_close(mqd),
+		      "unable to close message queue descriptor.\n");
+	pthread_exit(p1);
+	return NULL;
+}
+
+void test_mqueue(void)
+{
+	mqd_t mqd;
+	struct mq_attr attrs;
+	s32_t mode = 0777, flags = O_RDWR | O_CREAT, ret, i;
+	void *retval;
+	pthread_attr_t attr[N_THR];
+	pthread_t newthread[N_THR];
+
+	attrs.mq_msgsize = MESSAGE_SIZE;
+	attrs.mq_maxmsg = MESG_COUNT_PERMQ;
+
+	mqd = mq_open(queue, flags, mode, &attrs);
+
+	for (i = 0; i < N_THR; i++) {
+		/* Creating threads */
+		if (pthread_attr_init(&attr[i]) != 0) {
+			pthread_attr_destroy(&attr[i]);
+			zassert_false(pthread_attr_init(&attr[i]),
+				      "pthread attr init failed\n");
+		}
+		pthread_attr_setstack(&attr[i], &stacks[i][0], STACKSZ);
+
+		if (i % 2) {
+			ret = pthread_create(&newthread[i], &attr[i],
+					     sender_thread,
+					     (void *)i);
+		} else {
+			ret = pthread_create(&newthread[i], &attr[i],
+					     receiver_thread,
+					     (void *)i);
+		}
+
+		zassert_false(ret, "Not enough space to create new thread\n");
+		pthread_attr_destroy(&attr[i]);
+	}
+
+	usleep(10 * USEC_PER_MSEC);
+
+	for (i = 0; i < N_THR; i++) {
+		pthread_join(newthread[i], &retval);
+	}
+
+	zassert_false(mq_close(mqd),
+		      "unable to close message queue descriptor.\n");
+	zassert_false(mq_unlink(queue), "Not able to unlink Queue\n");
+	printk("Exiting the test.");
+}
+
+void test_main(void)
+{
+	ztest_test_suite(test_posix_mqueue, ztest_unit_test(test_mqueue));
```

Comment:
any reason why can't we run this in user mode?

Reply:
Hi @andrewboie 

Initially when I started the implementation of POSIX layer, all APIs (like memory alloc, memory pool) were not USER space compliant. So currently  I have not tested with userspace.

I have added a new test case for message queue API set.